### PR TITLE
feat: Add logic to allow headers to be used instead of cookies for sessions

### PR DESCRIPTION
**Pull request by nkshah2**
## Summary of change
- Updates the core SDK logic to allow using headers instead of cookies for session management
- Updates existing tests to account for the new logic
- Adds tests for header based auth

## Related issues
-

## Test Plan
All existing tests pass and new tests were added

## Documentation changes
...

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
   - Along with the associated array in `SuperTokensIOS/Classes/Version.swift`
- [ ] Changes to the version if needed
   - In `SuperTokensIOS/Classes/Version.swift`
   - In `SuperTokensIOS.podspec`
- [x] Had installed and ran the pre-commit hook
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Update FDI version

### CODE REVIEW

File: SuperTokensIOS/Classes/SuperTokens.swift [hunk 1]
```diff
@@ -40,16 +40,15 @@ public class SuperTokens {
     internal static func resetForTests() {
         FrontToken.removeToken()
         AntiCSRF.removeToken()
-        IdRefreshToken.removeToken()
         SuperTokens.isInitCalled = false
     }
```
Comment (rishabhpoddar) [inline SuperTokensIOS/Classes/SuperTokens.swift:44]: add the function to get access token

File: SuperTokensIOS/Classes/Utils.swift [hunk 2]
```diff
@@ -171,6 +211,159 @@ internal class Utils {
...
+            formatter.dateStyle = .medium
+            formatter.timeStyle = .medium
+            formatter.timeZone = TimeZone(abbreviation: "UTC")
+            
+            expires = formatter.string(from: expiresDate)
+        }
+        
+        let domain = SuperTokens.config!.sessionTokenBackendDomain
+        
+        getUserDefaults().set("\(name)=\(value);expires=\(expires);domain=\("");path=/;samesite=lax", forKey: "st-storage-item-\(name)")
```
Comment (rishabhpoddar) [inline SuperTokensIOS/Classes/Utils.swift]: just store all tokens to not have the cookie format. No point adding extra complexity.

File: testHelpers/testapp/Tests/config.swift [hunk 3]
```diff
@@ -310,14 +310,27 @@ class ConfigTests: XCTestCase {
             apiDomain: "example.com"
         );
         XCTAssertEqual(SuperTokens.refreshTokenUrl , "https://example.com/auth/session/refresh");
-        XCTAssertEqual(SuperTokens.config!.cookieDomain , nil);
+        XCTAssertEqual(SuperTokens.config!.sessionTokenBackendDomain , nil);
 
         SuperTokens.resetForTests()
         try SuperTokens.initialize(
```
Comment (rishabhpoddar) [inline testHelpers/testapp/Tests/config.swift:316]: add a test that makes sure that older sessions are migrated to newer one smoothly

File: SuperTokensIOS/Classes/Utils.swift [hunk 4]
```diff
@@ -171,6 +211,159 @@ internal class Utils {
...
+        if let frontToken: String = headerFields[SuperTokensConstants.frontTokenHeaderKey] {
+            FrontToken.setItem(frontToken: frontToken)
+        }
+        
+        if let antiCSRF: String = headerFields[SuperTokensConstants.antiCSRFHeaderKey] {
+            let localSessionState = Utils.getLocalSessionState()
+            
+            if localSessionState.status == .EXISTS {
+                AntiCSRF.setToken(antiCSRFToken: antiCSRF, associatedAccessTokenUpdate: localSessionState.lastAccessTokenUpdate!)
+            }
```
Comment (rishabhpoddar) [inline SuperTokensIOS/Classes/Utils.swift]: why is this happening? Ask @porcellus if you are not sure, and comment in the code why this is happening

File: testHelpers/testapp/Tests/config.swift [hunk 5]
```diff
@@ -310,14 +310,27 @@ class ConfigTests: XCTestCase {
...
             apiDomain: "example.com",
-            cookieDomain: "a.b.example.com"
+            sessionTokenBackendDomain: "a.b.example.com"
         );
```
Comment (rishabhpoddar) [inline testHelpers/testapp/Tests/config.swift:319]: add new tests for this SDK and all other mobile SDKs

File: testHelpers/testapp/Tests/config.swift [hunk 6]
```diff
@@ -310,14 +310,27 @@ class ConfigTests: XCTestCase {
             apiDomain: "example.com"
         );
         XCTAssertEqual(SuperTokens.refreshTokenUrl , "https://example.com/auth/session/refresh");
-        XCTAssertEqual(SuperTokens.config!.cookieDomain , nil);
+        XCTAssertEqual(SuperTokens.config!.sessionTokenBackendDomain , nil);
```
Comment (rishabhpoddar) [inline testHelpers/testapp/Tests/config.swift:314]: Change FDI version in all frontend / backend SDKs?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: SuperTokensIOS/Classes/Utils.swift (hunk 4 above)
Comment (nkshah2) [reply SuperTokensIOS/Classes/Utils.swift]: The website SDK needs it for typing, removed it from this SDK since it handles nil checks in this case

**Pull request merged**